core/src/main/scala/com/evolutiongaming/catshelper/ReadWriteRef.scala

```diff
@@ -0,0 +1,160 @@
+package com.evolutiongaming.catshelper
+
+import cats.Applicative
+import cats.effect.{Concurrent, Resource}
+import cats.effect.concurrent.Ref
+import cats.implicits._
+
+/**
+ * A mutable reference to `A` value with read-write lock semantics:
+ *  - multiple "read" operations are allowed as long as there's no "write" running.
+ *  - "write" operation is exclusive.
+ *  - NB: "read" can't be promoted to "write" yet. Attempting a "write" inside "read" will deadlock.
+ *
+ * Build it with [[ReadWriteRef.PartialApply.of `ReadWriteRef[F].of(a)`]] or
+ * [[ReadWriteRef.of `ReadWriteRef.of[F, A](a)`]].
+ *
+ * @note Both [[ReadWriteRef.read `read`]] and [[ReadWriteRef.write `write`]] are implemented as `Resource`
+ *       since it has clear mental model of something that may be "in use".
+ */
+trait ReadWriteRef[F[_], A] {
+
+  /**
+   * "Read" operation. Multiple reads may be in use simultaneously.
+   * `reads.use(…)` blocks semantically while [[write]] in in use.
+   */
+  def read: Resource[F, A]
+
+  /**
+   * "Write" operation. Requires exclusive access.
+   * `write.use(…)` blocks semantically while [[read]] or another [[write]] is in use.
+   */
+  def write: Resource[F, ReadWriteRef.Upd[F, A]]
+}
+
+object ReadWriteRef {
+  /**
+   * A "resource" of [[ReadWriteRef.write `write`]].
+   * `(A => F[A])` part lets you read current value and produce a new one, effectfully.
+   * `F[A]` in the return part lets you use new `A` in the outer context, e.g. return it from `use`.
+   */
+  trait Upd[F[_], A] extends ((A => F[A]) => F[A])
+
+  object Upd {
+    implicit final class SelfOps[F[_], A](val self: Upd[F, A]) extends AnyVal {
+      /** A shorthand for non-effectful updates. */
+      def plain(f: A => A)(implicit F: Applicative[F]): F[A] = self(a => F.pure(f(a)))
+
+      /** A shorthand for non-effectful update that ignores previous value. */
+      def const(a: A)(implicit F: Applicative[F]): F[A] = self(_ => F.pure(a))
```

Comment:
`def set` ?

Reply:
Maybe, but again, it's a method on an "updater" so it kinda reads "update with a constant". Still, `set` might fit just fine.

Reply:
Renamed in 6be0e96.